pkg/kv/kvserver/replica_circuit_breaker.go

```diff
@@ -61,12 +62,141 @@ type replicaCircuitBreaker struct {
+// them with the circuit breaker. If the breaker is already tripped, its error
+// is returned immediately and the caller should not continue processing the
+// request. Otherwise, the cancel function is invoked when the breaker trips.
+// The caller is provided with a token and signaller for use in a call to
+// Unregister upon request completion. That method also takes the error (if any)
+// resulting from the request to ensure that in the case of a tripped breaker,
+// the error reflects this fact.
+func (br *replicaCircuitBreaker) Register(
+	ctx context.Context, cancel func(),
+) (_token interface{}, _ signaller, _ error) {
+	brSig := br.Signal()
+
+	if brSig.C() == nil {
+		// Circuit breakers are disabled, so don't do any work registering the
+		// context. Unregister will know that we didn't since it checks the
+		// same brSig.
+		return ctx, brSig, nil
+	}
+
+	// TODO(tbg): we may want to exclude more requests from this check, or allow
+	// requests to exclude themselves from the check (via their header). This
+	// latter mechanism could also replace isCircuitBreakerProbe.
+	if isCircuitBreakerProbe(ctx) {
+		brSig = neverTripSignaller{}
+	}
+
+	// NB: it might be tempting to check the breaker error first to avoid the call
+	// to Set below if the breaker is tripped at this point. However, the ordering
+	// here, subtly, is required to avoid situations in which the cancel is still
+	// in the map despite the probe having shut down (in which case cancel will
+	// not be invoked until the probe is next triggered, which maybe "never").
+	//
+	// To see this, consider the case in which the breaker is initially not
+	// tripped when we check, but then trips immediately and has the probe fail
+	// (and terminate). Since the probe is in charge of cancelling all tracked
+	// requests, we must ensure that this probe sees our request. Adding the
+	// request prior to calling Signal() means that if we see an untripped
+	// breaker, no probe is running - consequently should the breaker then trip,
+	// it will observe our cancel, thus avoiding a leak. If we observe a tripped
+	// breaker, we also need to remove our own cancel, as the probe may already
+	// have passed the point at which it iterates through the cancels prior to us
+	// inserting it. The cancel may be invoked twice, but that's ok.
+	//
+	// See TestReplicaCircuitBreaker_NoCancelRace.
+	tok := br.cancels.Set(ctx, cancel)
+	if err := brSig.Err(); err != nil {
+		br.cancels.Del(tok)
+		cancel()
+		return nil, nil, err
+	}
+
+	return tok, brSig, nil
+}
+
+// Unregister releases a tracked cancel function upon request completion.
+// The error resulting from the request is passed in to allow decorating
+// it in case the breaker tripped while the request was in-flight.
+//
+// See Register.
+func (br *replicaCircuitBreaker) Unregister(
+	tok interface{}, sig signaller, pErr *roachpb.Error,
+) *roachpb.Error {
+	brErr := sig.Err()
```

Comment:
Is there a particular reason for calling this so early, rather than where it's used?

Reply:
Oops, there was a bit of code missing, to handle brErr == nil with an early bail-out after removing from the registry. Added & moved declaration as far down as possible.